### ISSUE TITLE
Video Captions Support

### ISSUE THREAD

**Issue by bmattb** [GoogleForCreators/web-stories-wp#4189]
## Feature Description

As part of expected functionality the editor should allow for the ability of the user to associate video subtitle files (VTT files) with video media assets.

**[Product Brief](https://docs.google.com/document/d/1EQ12sD4a1plmMeBu7YMWOdUSg0FeegB9qi2_OlV3Tp0/edit#heading=h.h3uoy4bqqh92)**

[Figma](https://www.figma.com/file/NHHMwIfxMnz39NxqkrFb7R/Stories---Post-Stable?node-id=1069%3A230257)

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

**Comment by bmattb** [comment]
@o-fernandez to detail the requirements and release for this please.

**Comment by o-fernandez** [comment]
To keep this very simple, let's do the following: 

- When selecting a video, show a new field within Design -> Accessibility that is "captions"
- This field should have a button/affordance for the user to start an upload, which would allow the user to select a file from their local files and upload it. 
- This captions file is then associated with the video and captions load when the video is playing, use caption/subtitle support in https://github.com/ampproject/amphtml/pull/13751 

Can @choumx @barklund @pbakaus @samitron7 please review this and see if it makes sense. Any other details you need me to spec out? I can find time to do it, but don't want to block on me or UX if eng can make progress without us.

**Comment by spacedmonkey** [comment]
There are two place that the caption upload button should be added. 

In the design panel here under the design panel under accessibility 

<img width="268" alt="Screenshot 2020-09-15 at 17 42 05" src="https://user-images.githubusercontent.com/237508/93239810-3591b100-f77b-11ea-9f1e-8fcc2e7bad27.png">

There is also the edit media dialog. 

<img width="548" alt="Screenshot 2020-09-15 at 17 42 19" src="https://user-images.githubusercontent.com/237508/93239872-48a48100-f77b-11ea-848f-0f8c11d7d1d9.png">

This video widget found in WordPress, has an option for caption uploads looks like

<img width="883" alt="Screenshot 2020-09-15 at 17 44 12" src="https://user-images.githubusercontent.com/237508/93240010-7be71000-f77b-11ea-910e-b41e40cb9b23.png">

<img width="962" alt="Screenshot 2020-09-15 at 17 44 20" src="https://user-images.githubusercontent.com/237508/93240052-8d301c80-f77b-11ea-9b72-e61751443eaf.png">

The upload button should open the WordPress uploader and be limited to only `.vtt` files. Once selected, users should be able to remove / replace caption files, if wrong file was selected.

**Comment by samitron7** [comment]
@o-fernandez This approach looks good to me. As for where we should add the cta

1. Add add caption button to the accessibility section.
2. Clicking on it to trigger local upload.
3. Once uploaded, we show the file name with the option to remove.
4. Remove will take the users to the "Add video caption" state.

![Screen Shot 2020-09-21 at 11 31 44 AM](https://user-images.githubusercontent.com/10889431/93806643-4cfaef00-fbfe-11ea-801b-80377ffe4c28.png)


@spacedmonkey How did you get to the video widget? I can't seem to trigger that on a regular post.

**Comment by spacedmonkey** [comment]
@samitron7 

Step to get that view.

1. Hover over appearance on left hand sidebar. Click on widget. 
2. Click on Video. Insert into a widget area. 
3. Upload a video (mp4 file)
4. Then click edit video.

![Screenshot 2020-09-15 at 19 12 24](https://user-images.githubusercontent.com/237508/93248199-72fc3b80-f787-11ea-8dab-a357e3f8f959.png)

**Comment by o-fernandez** [comment]
@spacedmonkey to take a look and see if we can do this ahead of 1.0 release cut tomorrow.

**Comment by samitron7** [comment]
@spacedmonkey I made edits to the mocks to show a separate area for captions. I need to create a more consistent pattern for features where you can either delete/replace but this should work for now if we need to get this out today.

**Comment by spacedmonkey** [comment]
I did some more research into subtitles and it’s more complex than I originally thought. For one, even gutenberg does not support uploading subtitles files at this time. See https://github.com/WordPress/gutenberg/issues/7673 . More importantly, WordPress itself doesn’t not support uploading .vtt files ( or even srt ) which is the subtitle file format. This changes the scope of the work but a lot and needs more thought than I think we have time for right now. I think the video widget example ( I screenshoted in the ticket ) is a weird outlier in WordPress, as I was unable to get it to work in my testing. As WordPress itself doesn’t support this, I think ( as sad as it makes me ) that we should bump to 1.1. If it is not part of the existing WordPress experience, it becomes a nice to have IMO.

<img width="437" alt="Screenshot 2020-09-21 at 20 11 09" src="https://user-images.githubusercontent.com/237508/93810475-a1c25800-fc46-11ea-947a-edf9de1e91ae.png">

**Comment by spacedmonkey** [comment]
While diving into this I noticed that track tags have a number of required attributes, see the [docs](https://www.w3schools.com/tags/tag_track.asp). 

Fields I would flag, 
- `srclang` which is a language code. This required specially if the subtitles are not in english. I believe we would need a dropdown of country codes. For users to select. 
- `label` this is a user defined label for the subtitles. 
Optional 
- `kind` Type of captions uploaded, we would default to `subtitles`. 

We would need fields to be able to input these values in the design panel. 

I have a WIP PR #4615 of what I have done so far. It need tests and more design work.

**Comment by bmattb** [comment]
@o-fernandez  / @samitron7  this requires further clarification before it can be actioned.

**Comment by o-fernandez** [comment]
@bmattb can we please add into the ticket what is the clarification needed? A handy summary of the open questions would be helpful. I read through the comments and it wasn't evident what needs product/UX clarification to unblock progress.

**Comment by swissspidy** [comment]
@o-fernandez Here's my summary:

* One or more subtitle [track](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/track) files can be added to a video
* For each file, you **should** declare **how the text track is meant to be used** (`kind` attribute):  
  - subtitles: Subtitles provide translation of content that cannot be understood by the viewer. For example dialogue or text that is not English in an English language film.
  - captions: Closed captions provide a transcription and possibly a translation of audio.
  - descriptions: Textual description of the video content.
  - chapters: Chapter titles are intended to be used when the user is navigating the media resource.
  - metadata: Tracks used by scripts. Not visible to the user.
* For each file, you **can** add a **label**, a user-readable title of the text track which is used by the browser when listing available text tracks.
* For each file, you **should** add the **language** of the track text data.

We need to:

* Add an option to specify the `kind` of track **or** at least make a clear decision on what we use as the default value
* Add a language option
* Decide whether we should  add a label option (maybe AMP doesn't use it?)
* create a design

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/track is a good read for this

**Comment by swissspidy** [comment]
Once we implement this, we need to make sure subtitles/captions are displayed properly on the frontend. Right now they can be hidden by overlapping items. Screenshot:

<img width="418" alt="Screenshot 2020-09-24 at 00 23 29" src="https://user-images.githubusercontent.com/841956/94075832-2dc9b080-fdfc-11ea-9b69-dd0d7adc0bec.png">

The [`::cue`](https://developer.mozilla.org/en-US/docs/Web/CSS/::cue) pseudo element might be useful here.

**Comment by o-fernandez** [comment]
For this ticket specifically, I'd suggest we focus on: 
- User can add a "captions" file (so captions `kind` in Pascal's list)
- There will be only one captions file per video, so no need for customizing a label.
- The *language* attribute should be set by the user among available options. 

Since we will want to add subtitles support as well in the future, good to consider the UX for uploading a captions file and its extensibility into other types. 

Would be good to consider the UX both in the editor and from the WP media gallery, but we can narrowly focus on the UX within the Editor for this release/ticket. 

@samitron7 let me know if this makes sense. We can chat about this. 

UPDATE: write a short **[product brief](https://docs.google.com/document/d/1EQ12sD4a1plmMeBu7YMWOdUSg0FeegB9qi2_OlV3Tp0/edit#heading=h.h3uoy4bqqh92)** for this

**Comment by samitron7** [comment]
Attached figma file in description. Also [here](https://www.figma.com/file/NHHMwIfxMnz39NxqkrFb7R/Stories---Post-Stable?node-id=1069%3A230257).

**Comment by bmattb** [comment]
@barklund  to review and point and confirm if this is feasible for S39

**Comment by spacedmonkey** [comment]
@samitron7 I don't see to have access to that link you shared. Can you give me access / fix the link please 😃

**Comment by swissspidy** [comment]
Took screenshots for now:

**Initial appearance:**

![Screenshot 2020-10-08 at 13 59 56](https://user-images.githubusercontent.com/841956/95455702-8fe2f380-096e-11eb-893e-be5f1fe6f583.png)

Clicking on the button opens the browser/OS file dialog to upload a file

**Uploading state**

![Screenshot 2020-10-08 at 14 03 48](https://user-images.githubusercontent.com/841956/95456097-1a2b5780-096f-11eb-911e-f02ed21b0b0e.png)

**Upload error**

Display error message if there is an upload error. After a short delay, remove the message.

![Screenshot 2020-10-08 at 14 02 18](https://user-images.githubusercontent.com/841956/95455951-e2241480-096e-11eb-8a27-4c0cb2c2adcd.png)

**Note:** Exact wording TBD

**Successful upload**

After uploading, the language can be chosen from a dropdown.

IIRC we discussed also allowing typing (like the font picker)

**Note:** exact way of how we get the list of locales is still **TBD** - for testing we can use a hardcoded list for now.

![Screenshot 2020-10-08 at 14 00 53](https://user-images.githubusercontent.com/841956/95455804-b143df80-096e-11eb-9b2f-731adde95418.png)

![Screenshot 2020-10-08 at 14 01 33](https://user-images.githubusercontent.com/841956/95455878-c7ea3680-096e-11eb-9b43-fa3a4252ff32.png)

**Multiple captions uploaded**

![Screenshot 2020-10-08 at 14 04 40](https://user-images.githubusercontent.com/841956/95456169-36c78f80-096f-11eb-9e22-0191e82580f8.png)

**Comment by spacedmonkey** [comment]
@swissspidy For the dropdown of languages, should we use the value found in WordPress. See Options > settings. 

<img width="253" alt="Screenshot 2020-10-09 at 13 01 36" src="https://user-images.githubusercontent.com/237508/95580847-ed8b4480-0a2f-11eb-96ac-e756afdfd2b9.png">

The data for this dropdown comes from [wp_get_available_translations](https://developer.wordpress.org/reference/functions/wp_get_available_translations/).

It outputs data in this format. 
```

Array
(
    [af] => Array
        (
            [language] => af
            [version] => 5.5.1
            [updated] => 2020-09-20 19:59:25
            [english_name] => Afrikaans
            [native_name] => Afrikaans
            [package] => https://downloads.wordpress.org/translation/core/5.5.1/af.zip
            [iso] => Array
                (
                    [1] => af
                    [2] => afr
                )

            [strings] => Array
                (
                    [continue] => Gaan voort
                )

        )

    [ar] => Array
        (
            [language] => ar
            [version] => 5.5.1
            [updated] => 2020-09-24 06:56:02
            [english_name] => Arabic
            [native_name] => العربية
            [package] => https://downloads.wordpress.org/translation/core/5.5.1/ar.zip
            [iso] => Array
                (
                    [1] => ar
                    [2] => ara
                )

            [strings] => Array
                (
                    [continue] => المتابعة
                )

        )
```

Or we could use [get_available_languages](https://developer.wordpress.org/reference/functions/get_available_languages/)? 

Some more questions
1. Should we be using native names or english names here? 
2. Should we highlight the current language? 
3. Should the dropdown be a search box, similar to the font picker / select2?
4. Once a language is selected, should it be removed dropdown so it can not be selected again?
5. From the design, there is an uploading progress meter. However in my POC PR, I use the WordPress media dialog filtered to ttf files. Is this acceptable? 
6. How will this data be stored on the story object?

Allowing  for multiple languages and language dropdown, add complexity. This would change the point number to more like 5/8 points.

**Comment by swissspidy** [comment]
Trying to add some answers where I can, but deferring the others to @o-fernandez and @samitron7 for clarification.

We've discussed some of it on the [product brief](https://docs.google.com/document/d/1EQ12sD4a1plmMeBu7YMWOdUSg0FeegB9qi2_OlV3Tp0/edit#heading=h.h3uoy4bqqh92), but that wasn't reflected in the brief itself and thus does not match the design :/

> 0. For the dropdown of languages, should we use the value found in WordPress. See Options > settings.

Not sure. As I mentioned in my comment above, the exact way of how we get the list of locales is still TBD - for testing we can use a hardcoded list for now.

It's probably easiest if we just bundle a list of locales in our plugin (like [GlotPress](https://github.com/GlotPress/GlotPress-WP/blob/develop/locales/locales.php) or [WordPress.org does](https://meta.trac.wordpress.org/browser/sites/trunk/wordpress.org/public_html/wp-content/mu-plugins/pub/locales/locales.php)) — but this needs a bit of exploration first.

Aside: `wp_get_available_translations` is fetched via an HTTP requests (and then cached).

> 1. Should we be using native names or english names here?

Native names! Otherwise a user will never find their language if they don't speak English :)

> 3. Should the dropdown be a search box, similar to the font picker / select2?

> 2. Should we highlight the current language?

Yes. Reusing the font picker would be good.

Citing the discussion on the product brief comments:

"Free-form search can work if they type in French and we show the proper name"

"Just as long as we show all the tags for german when you type german."

"Also, can we determine the country/locale and push recommended languages to the top of the drop down. Overall, I imagine the functionality to be close to what we have for font family now if we can do those 2 things."

> 4. Once a language is selected, should it be removed dropdown so it can not be selected again?

Good question! IMO I don't think we have to do that. Don't really see a reason for removing.

> 6. How will this data be stored on the story object?

Any suggestions? :-)

I assume they are simply part of the video element object (like `resource`). Maybe something like this is enough:

```json
{
  "tracks": [
    {
      "id": 123,
      "src": "https://exampe.com/file.vtt",
      "kind": "captions",
      "lang": "en"
    }
  ]
}

**Comment by spacedmonkey** [comment]
> Native names! Otherwise a user will never find their language if they don't speak English :)

I disagree on this one. If am an english speaker and I am given translations for say Spanish or German, I would search, German and Spanish. I would not to think to search for `Deutsch` or `Español`.  But then we have the problem of having to maintain a translated list. 

>Yes. Reusing the font picker would be good.

I seem to remember that this isn't that easy to do. But I could be wrong. Font picker is extremely custom at this point. 

> Aside: wp_get_available_translations is fetched via an HTTP requests (and then cached).

I know, which is why I want to avoid using it, as they is a performance reasons. 

The current implementation only has one track in mind. 

https://github.com/google/web-stories-wp/pull/4615/files#diff-82d994ff398776bb64dd06a3bcd44fbdR42-R43

It should be a simple tweak to convert this to an array of objects.

**Comment by o-fernandez** [comment]
Product brief updating is not my strength... But let me just add in my 2c here and will reflect on the brief: 
 
> > 1. Should we be using native names or english names here?
> 
> Native names! Otherwise a user will never find their language if they don't speak English :)

Could we show both for non-English, something like this? 
日本語 (Japanese)

And the user is able to find it by typing in any part of the string? 

> 
> > 1. Should the dropdown be a search box, similar to the font picker / select2?
> > 1. Should we highlight the current language?
> 
> Yes. Reusing the font picker would be good.

Yes, to both of these questions. We should use the picker and allow the user to input something to filter down the list. Ideally, we'd also have a recently used section like in the font picker. 

> 
> Citing the discussion on the product brief comments:
> 
> "Free-form search can work if they type in French and we show the proper name"
> 
> "Just as long as we show all the tags for german when you type german."
> 
> "Also, can we determine the country/locale and push recommended languages to the top of the drop down. Overall, I imagine the functionality to be close to what we have for font family now if we can do those 2 things."
> 
> > 1. Once a language is selected, should it be removed dropdown so it can not be selected again?
> 
> Good question! IMO I don't think we have to do that. Don't really see a reason for removing.

Agree! No need to do this.

**Comment by barklund** [comment]
> > 2. Should we highlight the current language?
> > 3. Should the dropdown be a search box, similar to the font picker / select2?
> 
> Yes. Reusing the font picker would be good.

Note that the font picker has not yet been generalized - it's a dropdown with three bonus features, that other places want to reuse:

- Seach
- Suggestions
- Preview

We have 3 potential use cases for it now:

- Language dropdown (with search and suggestions) [currently this ticket]
- Author dropdown (with search and maybe suggestions) [#4749]
- Font-weight dropdown (with preview) [no ticket]

I don't think this generalization job fits this ticket, as it's quite complex to do.

I would recommend to do it in a follow-up ticket, and only after #4749 has been completed (as it seems the most important one of the three to do).

> "Just as long as we show all the tags for german when you type german."

If we use native names, we'll show nothing when searching for "German" - only when searching for "Deutsch". Or do we want some custom search logic, that'll both search for either English or native name, but always show native name? Or both, as Omar suggests.

**Comment by spacedmonkey** [comment]
I think that we should do as Omar says, 

Deutsch (German). When a user searches, it search both native and english name. I think it is to hard to maintain a list of translations for all languages, so German in Spanish, Spanish in German etc. WordPress.org, as 201 languages, that is 201 x 201 to be translated. I think if a user does not know the native language, english is a good default, I would say,

**Comment by spacedmonkey** [comment]
I have been thinking about this data structure.
```
{
  "tracks": [
    {
      "id": 123,
      "src": "https://exampe.com/file.vtt",
      "kind": "captions",
      "lang": "en"
    }
  ]
}
```

As we have a repeatable field to add an many track files as you want, it could be possible to add the same file twice. So each object will need a unique id to reference. I will take a look into how pages have their page id generated.

**Comment by swissspidy** [comment]
> 5. From the design, there is an uploading progress meter. However in my POC PR, I use the WordPress media dialog filtered to `ttf` files. Is this acceptable?

@o-fernandez @samitron7 Thoughts on this? Perhaps that's OK for a first iteration, but we then customize it in a next step?

(Aside: I assume you meant `vtt` files)

----

Overall it we've uncovered quite a few unknowns which, as Jonny pointed out, increases complexity. Feels like we might not be able to get this done for the next release.

@spacedmonkey could you perhaps summarize your current findings and work a bit? What are your thoughts?

**Comment by o-fernandez** [comment]
Assuming it's vtt files, then yes, all else sounds OK as a starting point. Will wait for @spacedmonkey's update with the summary.

**Comment by spacedmonkey** [comment]
I am still working on the PR. I will update at the end of the day.

**Comment by spacedmonkey** [comment]
I have something working locally. It needs styling and tests to be added, but it functionally working. 

I just want to confirm with @samitron7 @o-fernandez that the UX ( please ignore the styling for now ). Here is a video of the UX. https://www.youtube.com/watch?v=5wtRyFt3UjU&feature=youtu.be

The important thing to note here is the use of WordPress media picker here. This picker is limited to VTT files and will not let you pick anything but these files. 

Before this can be merged, we need to work out the list of languages and how we are going to populate this date. I think there need to a fixed list for V1. But you will also need to add the users current language. For example if the fix list is german, english, french and Italian, the user should be able to pick hebrew as the language. 

@swissspidy Have you had any thoughts on the list of languages / selecting the current language.

If you want to play with my PR, it can be found here. #4615

**Comment by o-fernandez** [comment]
I'll let @samitron7 comment on the UX, I think this is acceptable with some refinements as p1/p2 to do later IMO. 

Re: the list of languages, seems like we need to allow for any of these? https://r12a.github.io/app-subtags/
There's quite a few of them, so I think that the language picker that is similar to the font picker (with an input field to filter the list) seems important. The drop-down can just have the most popular language codes as the first ones shown if the user doesn't search for anything. 

English (en)
Chinese (zh)
Hindi (hi)
Spanish (es) 
French (fr)
Arabic (ar)
Bengali (bn)
Russian (ru)
Portuguese (pt)
Indonesian (id)

We can have more, but I think that the user should be able to easily type a language (e.g., Italian) and get the right code. Then, with the "recently used" section of the picker they could see that next time they are adding a caption (if we can easily implement that part of the picker).

**Comment by bmattb** [comment]
recommendation for MVP (this ticket) - use the WP language list of ~100 languages with no preselected option
Future (1.2?) - use a list similar to fonts to provide auto-complete function

**Comment by spacedmonkey** [comment]
As agreed, I have implemented the language dropdown with all WordPress languages. The dropdown is extremely long and not specially usable. Is this okay for an MPV - https://youtu.be/OQFlz0izD_E ?

**Comment by bmattb** [comment]
@o-fernandez - in backlog grooming it was flagged that this MVP version of this will not be completed/ready for end of S39 and will take additional time in S40 to complete, **however** INFRA pod feels that they could prioritize this for the next week and still have it included in scope for v1.1

**Comment by barklund** [comment]
Note that both search and recently used items require that #4749 is implemented first. I have prioritized this for the next sprint, so @miina or I might even get to that this week.

The generalization of the font picker can happen without the API for users present in that ticket to be completed in two separate PRs.

**Comment by spacedmonkey** [comment]
I feel like the infra pod has done it's work here and this should now be handed off to the editor pod, to finish off. 
I think we should merge what we have now behind a feature flag ( after code review is complete ). Then create another ticket that would be style the inputs, add some more tests and implement #4749. Thoughts? 

Update: ticket created for styling https://github.com/google/web-stories-wp/issues/4899

**Comment by swissspidy** [comment]
As you mentioned there's #4899 to hand off styling to Prometheus/Pea pod. I'll update that ticket with details shortly.

If you can add the feature flag to #4615 that's fine with me. Then we can code-review & merge it.

The update on #4749 is promising. I think it warrants a new ticket to switch dropdown implementation for captions, I'll create that ticket.

**Comment by o-fernandez** [comment]
> As agreed, I have implemented the language dropdown with all WordPress languages. The dropdown is extremely long and not specially usable. Is this okay for an MPV - https://youtu.be/OQFlz0izD_E ?

I would be OK with this going out in 1.1 as it is, and prioritize the styling (#4899) for 1.2. In fact, if we can do a small release between 1.1 and 1.2 that includes #4899 all the better. @choumx @swissspidy

**Comment by swissspidy** [comment]
No need to overcomplicate things. Adding tests and improving styling (without autoomplete dropdown) shouldn't take more than a few hours

**Comment by spacedmonkey** [comment]
I have made some discoveries while finalizing the code. 

- The AMP stories format does not support track tag with type of subtitles. This is because video controls are disabled there is no way for a user to enable or change language, as the controls are not there to enable this. 
- The AMP stories format does support captions ( see [example](https://wp.stories.google/stories/creating-a-web-story-in-less-than-5-minutes-storytime/) ), but only one set of captions. Again, because the user can not select a different language, only one track is supported. 
- Captions should be enabled on video backgrounds. 

Some thoughts
- As there can only be one caption file, we should not allow them to upload multiples. Once user uploads one caption file then add more button should be removed / disabled. 
- As there is only one caption file, we could remove the language option for now. Language is [only required](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/track#attr-srclang) for tracks defined as subtitles. We can default the language to the users language for now. Captions are text of the spoken words, in the language they are spoken. It is likely that users language will match the video's language. 
- All references to subtitles should be removed from designs and code, as it is just captions for now. 
- Design may need to be updated if only on caption file is allowed. 

@o-fernandez Thoughts. The above sound reasonable?

**Comment by spacedmonkey** [comment]
[Example](https://playground.amp.dev/?format=websites#share=PCFkb2N0eXBlIGh0bWw+CjxodG1sIOKaoT4KPGhlYWQ+CiAgPG1ldGEgY2hhcnNldD0idXRmLTgiPgogIDx0aXRsZT5hbXAtdmlkZW8gLSBFeGFtcGxlIDE8L3RpdGxlPgogIDxzY3JpcHQgYXN5bmMgc3JjPSJodHRwczovL2Nkbi5hbXBwcm9qZWN0Lm9yZy92MC5qcyI+PC9zY3JpcHQ+CiAgPHNjcmlwdCBjdXN0b20tZWxlbWVudD0iYW1wLXZpZGVvIiBzcmM9Imh0dHBzOi8vY2RuLmFtcHByb2plY3Qub3JnL3YwL2FtcC12aWRlby0wLjEuanMiIGFzeW5jPjwvc2NyaXB0PgogIDxsaW5rIHJlbD0iY2Fub25pY2FsIiBocmVmPSJodHRwczovL3ByZXZpZXcuYW1wLmRldi9kb2N1bWVudGF0aW9uL2NvbXBvbmVudHMvYW1wLXZpZGVvLmV4YW1wbGUuMS5odG1sIj4KICA8bWV0YSBuYW1lPSJ2aWV3cG9ydCIgY29udGVudD0id2lkdGg9ZGV2aWNlLXdpZHRoLG1pbmltdW0tc2NhbGU9MSxpbml0aWFsLXNjYWxlPTEiPgogIDxzdHlsZSBhbXAtYm9pbGVycGxhdGU+Ym9keXstd2Via2l0LWFuaW1hdGlvbjotYW1wLXN0YXJ0IDhzIHN0ZXBzKDEsZW5kKSAwcyAxIG5vcm1hbCBib3RoOy1tb3otYW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGg7LW1zLWFuaW1hdGlvbjotYW1wLXN0YXJ0IDhzIHN0ZXBzKDEsZW5kKSAwcyAxIG5vcm1hbCBib3RoO2FuaW1hdGlvbjotYW1wLXN0YXJ0IDhzIHN0ZXBzKDEsZW5kKSAwcyAxIG5vcm1hbCBib3RofUAtd2Via2l0LWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1ALW1vei1rZXlmcmFtZXMgLWFtcC1zdGFydHtmcm9te3Zpc2liaWxpdHk6aGlkZGVufXRve3Zpc2liaWxpdHk6dmlzaWJsZX19QC1tcy1rZXlmcmFtZXMgLWFtcC1zdGFydHtmcm9te3Zpc2liaWxpdHk6aGlkZGVufXRve3Zpc2liaWxpdHk6dmlzaWJsZX19QC1vLWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1Aa2V5ZnJhbWVzIC1hbXAtc3RhcnR7ZnJvbXt2aXNpYmlsaXR5OmhpZGRlbn10b3t2aXNpYmlsaXR5OnZpc2libGV9fTwvc3R5bGU+PG5vc2NyaXB0PjxzdHlsZSBhbXAtYm9pbGVycGxhdGU+Ym9keXstd2Via2l0LWFuaW1hdGlvbjpub25lOy1tb3otYW5pbWF0aW9uOm5vbmU7LW1zLWFuaW1hdGlvbjpub25lO2FuaW1hdGlvbjpub25lfTwvc3R5bGU+PC9ub3NjcmlwdD4KPC9oZWFkPgo8Ym9keT4KCjxhbXAtdmlkZW8gY29udHJvbHMgY3Jvc3NvcmlnaW4KICB3aWR0aD0iNjQwIgogIGhlaWdodD0iMzYwIgogIGxheW91dD0icmVzcG9uc2l2ZSIKICBwb3N0ZXI9Ii9zdGF0aWMvaW5saW5lLWV4YW1wbGVzL2ltYWdlcy9raXR0ZW4tcGxheWluZy5wbmciPgogIDxzb3VyY2Ugc3JjPSIvc3RhdGljL2lubGluZS1leGFtcGxlcy92aWRlb3Mva2l0dGVuLXBsYXlpbmcud2VibSIKICAgIHR5cGU9InZpZGVvL3dlYm0iIC8+CiAgPHNvdXJjZSBzcmM9Ii9zdGF0aWMvaW5saW5lLWV4YW1wbGVzL3ZpZGVvcy9raXR0ZW4tcGxheWluZy5tcDQiCiAgICB0eXBlPSJ2aWRlby9tcDQiIC8+CiAgPHRyYWNrIHNyY2xhbmc9ImVuLWNhIiBsYWJlbD0iRW5nbGlzaCAoQ2FuYWRhKSIga2luZD0ic3VidGl0bGVzIiBzcmM9Imh0dHBzOi8vZ2lzdC5naXRodWJ1c2VyY29udGVudC5jb20vc2FtZHV0dG9uL2NhMzdmM2FkYWY0ZTIzNjc5OTU3YjgwODNlMDYxMTc3L3Jhdy9lMTkzOTlmYmNjYmMwNjlhMmFmNDI2NmU1MTIwYWU2YmFkNjI2OTlhL3NhbXBsZS52dHQiPjwvdHJhY2s+CiAgPGRpdiBmYWxsYmFjaz4KICAgIDxwPlRoaXMgYnJvd3NlciBkb2VzIG5vdCBzdXBwb3J0IHRoZSB2aWRlbyBlbGVtZW50LjwvcD4KICA8L2Rpdj4KPC9hbXAtdmlkZW8+CjwvYm9keT4KPC9odG1sPg==) of subtitles in amp playground.

**Comment by swissspidy** [comment]
> The AMP stories format does not support track tag with type of subtitles. This is because video controls are disabled there is no way for a user to enable or change language, as the controls are not there to enable this.

Not 100% accurate. Subtitles _do_ work, at least they did for me in AMP playground. The `default` track always works, even with `subtitles`:

```html
<track default srclang="en" label="English" kind="subtitles" src="https://example.com/sample.vtt"></track>
```

But given the fact that `captions` are P1 and `subtitles` are P2, I agree with the suggestion above— just going with `captions` for now to keep things simple (this drastically reduces complexity!)

So the next steps are:

* Remove languages logic
* Default to `captions`
* Remove UI to add multiple tracks\*

\* `element.tracks` should still be an array, in case we need it in the future

After that, we can improve styling in #4899

**Comment by spacedmonkey** [comment]
This ticket is ready for code review. Now the ticket is much more simple, I think we can get this merge and tested for 1.1.

**Comment by csossi** [comment]
Verified in QA
